Add descriptive text box for variables, reorder variables in dropdown

nicolejkeeney opened the pull request:
### This PR accomplishes two main goals: 

1. It adds a descriptive box below the variable selector dropdown with a layman's description of the variable selected. The text dynamically updates with new selections. See screenshot below for an idea of what this looks like (the added text is circled). 
2. It reorders the variables in the drop down selector to match the order of variables in the csv file containing the extended variable descriptions. **Therefore, in order to re-order the variables, you just change the order of the variables in the csv file** (no need to change the source code). 

<img width="500" alt="extended_var_screenshot" src="https://user-images.githubusercontent.com/66140951/178777147-03f12a34-496e-4654-89ff-663338020770.png">

### Notable code/data additions:
1. A new file **variable_descriptions.csv** that contains each variable name, the catalog description (from whatever model used to generate it), and an extended variable description. This file is located in climakitae/data (a new folder). 
2. A new unit testing file t**ests/test_var_csv.py** to ensure that the variable descriptions csv file is compatible with the variables in the catalog, to avoid unexpected issues in the source code if the csv file is changed. As in, we need the variable names and catalog descriptions of each variable in the catalog and the csv file to match to avoid conflicts. 

### Important note: 
One of the tests in the new testing file is not currently passing. This is due to a bug in main (the variable "Precipitation (total)" is lacking a variable name). This bug will be fixed once the add_var_test branch is merged. 

EDIT 07/22/2022: The tests all pass after merging the add_var_test branch with main!

elehmer commented on climakitae/data/variable_descriptions.csv:
```diff
@@ -0,0 +1,22 @@
+SNOWNC,Snowfall (snow and ice),Total precipitation in the form of snow and ice. 
+Q2,2m Water Vapor Mixing Ratio,"The ratio of the mass of water vapor over the mass of dry air, measured 2 meters above Earth's surface."
+TSK,Surface skin temperature,Temperature of the Earth's surface (ground surface).
+U10,West-East component of Wind at 10m,"Wind coming from the west, measured at 10 meters above Earth's surface. By covention, wind coming from the east is negative.  "
```
Extra white space at end of description.

elehmer commented on climakitae/data/variable_descriptions.csv:
```diff
@@ -0,0 +1,22 @@
+LWUPBC,Instantaneous upwelling clear sky longwave flux at bottom,Longwave radiation emitted by Earth's surface that passes through the atmosphere (i.e. is not reflected back to the surface or absorbed by the atmosphere). 
+SWDNB,Instantaneous downwelling shortwave flux at bottom,"Shortwave radiation from the sun that reaches Earth's surface, minus radiation reflected back to space by clouds or absorbed by the atmosphere. "
+SWDNBC,Instantaneous downwelling clear sky shortwave flux at bottom,Shortwave radiation from the sun that reaches Earth's surface. 
+SWUPB,Instantaneous upwelling shortwave flux at bottom,"Shortwave radiation from the sun that is reflected by Earth's surface, some of which may be reflected back downwards by clouds (i.e. not reflected back to space). "
```
Extra white space at end of description.

elehmer reviewed (approved):
Looks good, but if you could clean up the CSV file that would be great (see comments there). Thanks!

elehmer commented:
Does the CSV work with the quotations removed? Quotations are needed when there are "," in the field as the comma is used as the field delimiter. So the quotes might need to go back?

elehmer commented:
> > Does the CSV work with the quotations removed? Quotations are needed when there are "," in the field as the comma is used as the field delimiter. So the quotes might need to go back?
> 
> No it doesn't work, I'm working on adding the quotations back in now. I got confused by your comment and made a mess out of the file.

Sorry about that! Yeah, just need to clean up the white space!

nicolejkeeney commented:
>

nicolejkeeney commented:
Ugh, me and GitHub are not friends today. I accidentally clicked "Close with comment" to this PR. I just reopened it. Sorry for any confusion.

ckingdon95 commented on tests/test_var_csv.py:
```diff
@@ -0,0 +1,36 @@
+
+def test_csv_matches_cat_contents(cat_contents, descrip_dict_formatted):
+    """Ensure that items in catalog are the same as (or a smaller subset of) the items in the variable csv. """
+    assert cat_contents["hourly"]["Dynamical"].items() <= descrip_dict_formatted.items(), "Catalog contents are not the same as (or a smaller subset of) the items in the variable descriptions csv."
```
Hmm I just tried running this locally and this test failed for me. The first thing I noticed is one item in `cat_contents["hourly"]["Dynamical"]` is `'Precipitation (total)': ''`
whereas in the `descrip_dict_formatted` it is `'Precipitation (total)':'TOT_PRECIP'`
Happy to look into this more if it would be helpful, but you know this code better so will probably know what this is right away?

nicolejkeeney replied:
Yeah this test doesn't pass at the moment. This is due to a bug in main (the variable "Precipitation (total)" is lacking a variable name). This bug will be fixed once the add_var_test branch is merged, and the test will pass! Kind of cool that the unit test catches the bug, though :D

ckingdon95 replied:
great!

odoherty replied:
suggested order of variables:

2m temp
ppt
rh
wind_mag
wind_dir
u10
v10
PSFC
SWDIFF

maybe put the precip components last?

after that doesn't matter

odoherty replied:
nice job on the radiation description wording - looks like these are proper atmospheric science definitions.  suggest a cross-check with WRF handbook to ensure that the model world is doing the same thing as is expected (especially the comment about clouds catching reflected radiation - want to make sure this is true in WRF)

nicolejkeeney replied:
The tests are all passing now!!!!! WOOHOOOO!!!

odoherty commented on climakitae/data/variable_descriptions.csv:
```diff
@@ -0,0 +1,22 @@
+name,description,extended_description
+T2,2m Air Temperature,"Temperature of the air 2 meters above Earth's surface. This is the measure of air temperature used for most modeling applications."
```
Suggest rename Air Temperature at 2m for consistency

nicolejkeeney replied:
changed!

odoherty commented on climakitae/data/variable_descriptions.csv:
```diff
@@ -0,0 +1,22 @@
+name,description,extended_description
+T2,2m Air Temperature,"Temperature of the air 2 meters above Earth's surface. This is the measure of air temperature used for most modeling applications."
+TOT_PRECIP,Precipitation (total),"Total precipitation. Computed by summing total gridscale precipitation and total cumulus precipitation."
```
I think "Precipitation" without the total would suffice for description.

nicolejkeeney replied:
@thenaomig @vicford Do either of you have opinions on this? I'm inclined to leave "total" in this description for clarity's sake (since we have the other precip variables as options) but don't feel strongly.

odoherty commented on climakitae/data/variable_descriptions.csv:
```diff
@@ -0,0 +1,22 @@
+name,description,extended_description
+T2,2m Air Temperature,"Temperature of the air 2 meters above Earth's surface. This is the measure of air temperature used for most modeling applications."
+TOT_PRECIP,Precipitation (total),"Total precipitation. Computed by summing total gridscale precipitation and total cumulus precipitation."
+REL_HUMIDITY,Relative Humidity,"The amount of water vapor (i.e. moisture) in the air, relative to the amount of water vapor needed to saturate the air at the same air temperature and pressure. Expressed as a percentage."
```
The percentage of water vapor (i.e. moisture) relative to the amount of water vapor needed to saturate the air at the current air temperature and pressure.

nicolejkeeney replied:
Fixed!

nicolejkeeney commented:
I deleted a few comments from Owen that were just repeats. They were attached to Cora's PR likely by mistake. The comments are the same as those he posted in his review, so I deleted the repeat comments  attached to Cora's PR for clarity.

thenaomig commented on climakitae/data/variable_descriptions.csv:
```diff
@@ -0,0 +1,22 @@
+SWUPB,Instantaneous upwelling shortwave flux at bottom,"Shortwave radiation from the sun that is reflected by Earth's surface, some of which may be reflected back downwards by clouds (i.e. not reflected back to space)."
+SWUPBC,Instantaneous upwelling clear sky shortwave flux at bottom,Shortwave radiation from the sun that is reflected by Earth's surface back to space.
+RAINC,Accumulated total cumulus precipitation,"Any form of precipitation (rain, snow, etc.) that is produced specifically by cumulus/cumulonimbus clouds. These are generally small clouds that bring heavy precipitation."
+RAINNC,Accumulated total grid scale precipitation,"Any form of precipitation (rain, snow, etc.) captured by the model resolution (i.e. 45km). Small clouds (including cumulus clouds) typically are excluded from this."
```
Suggesting the following for even greater clarity and some parallelism between these options:
TOT_PRECIP, Precipitation (total), ...
RAINC, Precipitation (cumulus portion only), "Accumulated total cumulus precipitation. Any form..."
RAINNC, Precipitation (grid-scale portion only), "Accumulated total grid scale precipitation. Any form..."

keeping the latter two towards the bottom.

Also, change the order so that SWDDIF moves down near the other SW radiation variables.

nicolejkeeney replied:
Awesome, I like these changes a lot. Thanks for the feedback. Just pushed those changes. RAINC and RAINCC are currently at the bottom of the dropdown. Let me know if you think they need to be moved up higher.
<img width="472" alt="Screen Shot 2022-07-22 at 4 15 08 PM" src="https://user-images.githubusercontent.com/66140951/180579599-8015d0fd-2470-424a-a7d9-411097586740.png">

vicford commented on climakitae/data/variable_descriptions.csv:
```diff
@@ -0,0 +1,22 @@
+U10,West-East component of Wind at 10m,"Wind coming from the west, measured at 10 meters above Earth's surface. By covention, wind coming from the east is negative."
+V10,North-South component of Wind at 10m,"Wind coming from the north, measured at 10 meters above Earth's surface. By convention, wind coming from the south is negative."
+PSFC,Surface Pressure,Air pressure at Earth's surface.
+SNOWNC,Snowfall (snow and ice),Total precipitation in the form of snow and ice. 
```
For clarity and consistency with `Precipitation (total)`, I'd recommend changing the `Snowfall (snow and ice)` description to "Portion of the total precipitation in the form of snow and ice."

Additionally, the order of snowfall could be moved downwards in the list to reside above RAINC and RAINNC, but that is minor and arguably just an opinion, and could stay as is re: order.

nicolejkeeney replied:
Agreed! Thanks for your input! :D

nicolejkeeney commented:
Owen had a comment about ensuring that the radiation variable descriptions are accurate in regards to WRF's treatment of radiation. In an effort to move forward with this PR, I'll plan to address that comment in a later PR if the descriptions require further editing after looking more into the WRF documentation.
